[EDR Workflows] [Osquery] Fix results pagination beyond 10,000 results

tomsonpl opened the pull request:
### Problem

Elasticsearch has a hard limit of 10,000 results when using offset-based pagination (`from` + `size`). This caused osquery live query results to be inaccessible beyond page 100 (with 100 results per page). Large-scale deployments with thousands of agents frequently exceed this limit.

### Solution

Implement hybrid pagination that automatically switches between:
- **Offset pagination** for results < 10,000 (fast, simple)
- **PIT + search_after pagination** for results ≥ 10,000 (overcomes ES limit)

### Key changes

- Add server-side PIT (Point in Time) search utility with lazy initialization
- Update `get_live_query_results` route to detect when deep pagination is needed and batch through results to find cursor position
- Add `close_pit` endpoint for resource cleanup
- Update client-side `useAllResults` hook to track PIT state and search_after values per page
- Add loading overlay when initializing deep pagination

### API changes

New optional query parameters for `/internal/osquery/live_queries/{id}/results`:
- `pitId` - PIT identifier for pagination continuation
- `searchAfter` - JSON-encoded sort values for next page

New response fields:
- `pitId` - Returned PIT ID for subsequent requests
- `searchAfter` - Cursor for next page
- `hasMore` - Boolean indicating more results exist

New internal endpoint:
- `POST /internal/osquery/live_queries/pit/close` - Cleanup PIT resources


### Result 

https://github.com/user-attachments/assets/7d2fc382-8ec5-4d0a-8fe3-00da0339fbb6

tomsonpl committed:
feat(osquery): add PIT pagination constants and types

Add constants for Point in Time (PIT) pagination to support querying
beyond Elasticsearch's 10k offset limit.

Update ResultsStrategyResponse and ResultsRequestOptions types to
support pitId, searchAfter, and hasMore fields for cursor-based
pagination.

tomsonpl committed:
docs(osquery): add OpenAPI schemas for PIT pagination

Add OpenAPI schema definitions for PIT-based pagination parameters
and update io-ts route schema with pitId and searchAfter parameters.

Add close_pit.schema.yaml for internal PIT cleanup endpoint.

tomsonpl committed:
feat(osquery): implement PIT search utility and DSL builder

Add executePitSearch utility for executing Elasticsearch PIT searches
and update buildResultsQuery DSL builder for dual-mode pagination.

Add pagination utility functions in filters.ts.

Includes comprehensive test coverage for both modes.

tomsonpl committed:
feat(osquery): integrate PIT support into search strategy factory

Update allResults factory to support PIT-based pagination and
extract cursor data from ES response.

Update osquerySearchStrategyProvider to pass through pitId and
searchAfter parameters to the factory DSL builder.

Includes test coverage for the results factory parse function.

tomsonpl committed:
feat(osquery): add PIT pagination routes for live query results

Enhance get_live_query_results_route with hybrid pagination that
auto-detects when PIT mode is needed (offset >= 10k).

Add isSortResults type guard for safe searchAfter parsing.

Add close_pit_route internal endpoint for resource cleanup.

Includes comprehensive test mocks and test coverage.

tomsonpl committed:
feat(osquery): add client-side PIT pagination support

Update useAllResults hook for cursor-based pagination with state
tracking per page. Close PIT on unmount and state reset.

Add isFetchingWithoutCursor state for UX during initial deep page load.

Update ResultsTable with loading spinner overlay during cursor init.

Includes comprehensive test coverage for useAllResults hook.

tomsonpl committed:
refactor(osquery): rename cursor terminology to PIT pagination

Replace misleading "cursor" terminology with accurate Elasticsearch
PIT + search_after naming.

Aligns naming with Elasticsearch documentation and improves code clarity.

kibanamachine committed:
Changes from yarn openapi:bundle

kibanamachine committed:
Changes from yarn openapi:generate

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

tomsonpl committed:
refactor(osquery): memoize sort array and simplify useEffect

Memoize sort array in ResultsTable to prevent unnecessary re-renders.
Simplify PIT state reset effect by removing manual dependency comparison.

tomsonpl committed:
Merge remote-tracking branch 'origin/osquery-results-pit-search' into osquery-results-pit-search

kibanamachine committed:
Changes from make api-docs

tomsonpl committed:
fix(osquery): improve PIT pagination with abort signal and hasMore calculation

- Add abort signal support to PIT search for proper request cancellation
- Fix hasMore calculation to correctly determine if more pages exist
- Handle out-of-range page requests by returning empty page instead of
  falling back to offset pagination (which is capped at 10k)

tomsonpl committed:
Merge remote-tracking branch 'origin/osquery-results-pit-search' into osquery-results-pit-search

tomsonpl committed:
Merge branch 'main' into osquery-results-pit-search

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372344)
* Commit: 9d5d01aa1c6cc41e1dd1a8199da79ecaceb232db

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/372344#019b125d-b120-413b-817c-cef6cccbcaea)

### History
* :broken_heart: [Build #372301](https://buildkite.com/elastic/kibana-pull-request/builds/372301) failed a83f8c1c622d65416d985fb5417b0501ac631472
* :broken_heart: [Build #372269](https://buildkite.com/elastic/kibana-pull-request/builds/372269) failed 19a1922d00643a4a944157f4b640fe3793cc863e

cc @tomsonpl

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372344","number":372344,"commit":"9d5d01aa1c6cc41e1dd1a8199da79ecaceb232db"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372301","number":372301,"commit":"a83f8c1c622d65416d985fb5417b0501ac631472"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372269","number":372269,"commit":"19a1922d00643a4a944157f4b640fe3793cc863e"}],"number":372344}
buildkite-pr-comment-->

tomsonpl committed:
fix(osquery): remove error response schemas to fix OAS validation

Removes inline 400, 404, 500 error response definitions from the
get_live_query_results endpoint schema to prevent OAS validation
error count from increasing.

tomsonpl committed:
Merge remote-tracking branch 'origin/osquery-results-pit-search' into osquery-results-pit-search

kibanamachine committed:
Changes from make api-docs

tomsonpl committed:
fix(osquery): add missing query param descriptions

OAS validation requires parameter.description; add descriptions for PIT pagination params to prevent baseline error count increases.

tomsonpl committed:
Merge remote-tracking branch 'origin/osquery-results-pit-search' into osquery-results-pit-search

kibanamachine committed:
Changes from make api-docs

tomsonpl committed:
fix(osquery): prevent infinite re-render loop in useAllResults hook

Use a memoized resetKey string instead of depending directly on the
sort array in useEffect, preventing unnecessary re-renders when the
array reference changes but content remains the same.

tomsonpl committed:
Merge remote-tracking branch 'origin/osquery-results-pit-search' into osquery-results-pit-search

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372576)
* Commit: faaa106f4d326e8683a966a9c1ef09343a841069

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/372576#019b136e-fbfa-4185-90ce-042a08eb2bbb)

### History
* :broken_heart: [Build #372421](https://buildkite.com/elastic/kibana-pull-request/builds/372421) failed 530c325d2ab5ac2d795473e9855488c0b6146d03
* :broken_heart: [Build #372406](https://buildkite.com/elastic/kibana-pull-request/builds/372406) failed 1e74e7ccd2fdc94616d6dd23cb90b88ffdd2e2ef
* :broken_heart: [Build #372378](https://buildkite.com/elastic/kibana-pull-request/builds/372378) failed b421f468aba4b1f73915e9b2828950044a57b2c0
* :broken_heart: [Build #372368](https://buildkite.com/elastic/kibana-pull-request/builds/372368) failed c0ec9890227d9049d5fe58db384124d1c47878a9
* :broken_heart: [Build #372344](https://buildkite.com/elastic/kibana-pull-request/builds/372344) failed 9d5d01aa1c6cc41e1dd1a8199da79ecaceb232db

cc @tomsonpl

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372576","number":372576,"commit":"faaa106f4d326e8683a966a9c1ef09343a841069"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372421","number":372421,"commit":"530c325d2ab5ac2d795473e9855488c0b6146d03"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372406","number":372406,"commit":"1e74e7ccd2fdc94616d6dd23cb90b88ffdd2e2ef"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372378","number":372378,"commit":"b421f468aba4b1f73915e9b2828950044a57b2c0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372368","number":372368,"commit":"c0ec9890227d9049d5fe58db384124d1c47878a9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372344","number":372344,"commit":"9d5d01aa1c6cc41e1dd1a8199da79ecaceb232db"}],"number":372576}
buildkite-pr-comment-->

tomsonpl committed:
fix(osquery): fix max line length lint error

tomsonpl committed:
Merge remote-tracking branch 'origin/osquery-results-pit-search' into osquery-results-pit-search

florent-leborgne reviewed (approved):
LGTM from a docs perspective

tomsonpl committed:
fix(osquery): fix TypeScript error in useAllResults test

tomsonpl committed:
Merge branch 'main' into osquery-results-pit-search
